src/main/java/io/fusionauth/samlv2/service/DefaultSAMLv2Service.java

```diff
@@ -334,8 +334,8 @@ public String buildAuthnResponse(AuthenticationResponse response, boolean sign,
       document = encryptAssertion(document, encryptionAlgorithm, keyLocation, transportAlgorithm, encryptionCertificate, digest, mgf);
     }
 
-    // Sign the response if requested
-    if (sign && signatureOption == SignatureLocation.Response) {
+    // Sign the response if requested or if the response is a failure (successes may have had the assertion signed above)
+    if (sign && (signatureOption == SignatureLocation.Response || response.status.code != ResponseStatus.Success)) {
```

Comment:
Not sure on the context.

The origins of this library pre-date FusionAuth. We initially wrote what became this library for CleanSpeak so we could authenticate moderators against a their corp SAML v2 IdP. 

I haven't found anything that explicitly says you don't have to sign it when a signature is requested, so I think it is safe to assume we should not skip signing just because we failed for some other reason. 

The receiver will still want to be able to verify the authenticity of the response even if the response indicates we failed for some reason. 

I think it is wise to leave this existing behavior in place since it seems to be working so far.